File: backend/src/api/v1/notes.py
import uuid

from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.ext.asyncio import AsyncSession

from src.db.database import get_db
from src.notes.schema import (
    NoteCreate,
    NoteUpdate,
    NoteResponse,
)
from src.notes.service import note_service
from src.users.dependencies import get_current_user
from src.users.model import User

router = APIRouter(
    prefix="/projects",
    tags=["Notes"],
)


@router.post(
    "/{project_id}/notes",
    response_model=NoteResponse,
    status_code=status.HTTP_201_CREATED,
)
async def create_note(
    project_id: uuid.UUID,
    note_data: NoteCreate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return await note_service.create_note(
            db,
            project_id,
            note_data,
        )

    except ValueError as e:
        raise HTTPException(
            status_code=400,
            detail=str(e),
        )


@router.get(
    "/{project_id}/notes",
    response_model=list[NoteResponse],
)
async def get_notes(
    project_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    return await note_service.get_notes(
        db,
        project_id,
    )


@router.get(
    "/notes/{note_id}",
    response_model=NoteResponse,
)
async def get_note(
    note_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return await note_service.get_note(
            db,
            note_id,
        )

    except ValueError as e:
        raise HTTPException(
            status_code=404,
            detail=str(e),
        )


@router.patch(
    "/notes/{note_id}",
    response_model=NoteResponse,
)
async def update_note(
    note_id: uuid.UUID,
    note_data: NoteUpdate,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        return await note_service.update_note(
            db,
            note_id,
            note_data,
        )

    except ValueError as e:
        raise HTTPException(
            status_code=404,
            detail=str(e),
        )


@router.delete(
    "/notes/{note_id}",
    status_code=status.HTTP_204_NO_CONTENT,
)
async def delete_note(
    note_id: uuid.UUID,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
):
    try:
        await note_service.delete_note(
            db,
            note_id,
        )

    except ValueError as e:
        raise HTTPException(
            status_code=404,
            detail=str(e),
        )

File: frontend/src/lib/notes.ts
import api from "@/lib/api";
import { Note } from "@/types";

export async function getNotes(projectId: string): Promise<Note[]> {
  const response = await api.get(`/projects/${projectId}/notes`);

  return response.data.map((note: any) => ({
    id: note.id,
    projectId: note.project_id,
    title: note.title,
    content: note.content,
    updatedAt: note.updated_at,
  }));
}

export async function getNote(id: string): Promise<Note> {
  const response = await api.get(`/projects/notes/${id}`);

  const note = response.data;

  return {
    id: note.id,
    projectId: note.project_id,
    title: note.title,
    content: note.content,
    updatedAt: note.updated_at,
    createdAt: note.created_at,
  };
}

export async function createNote(
  projectId: string,
  data: {
    title: string;
    content: string;
  }
): Promise<Note> {
  const response = await api.post(
    `/projects/${projectId}/notes`,
    data
  );

  const note = response.data;

  return {
    id: note.id,
    projectId: note.project_id,
    title: note.title,
    content: note.content,
    updatedAt: note.updated_at,
    createdAt: note.created_at,
  };
}

export async function updateNote(
  id: string,
  data: {
    title?: string;
    content?: string;
  }
): Promise<Note> {
  const response = await api.patch(
    `/projects/notes/${id}`,
    data
  );

  const note = response.data;

  return {
    id: note.id,
    projectId: note.project_id,
    title: note.title,
    content: note.content,
    createdAt: note.created_at,
    updatedAt: note.updated_at,
  };
}

export async function deleteNote(id: string): Promise<void> {
  await api.delete(`/projects/notes/${id}`);
}
